src/stash/src/postgres.rs

```diff
@@ -114,17 +156,64 @@ impl Postgres {
     /// ```
     async fn transact<F, T>(&mut self, f: F) -> Result<T, StashError>
     where
-        F: for<'a> FnOnce(&'a mut Transaction) -> BoxFuture<'a, Result<T, StashError>>,
+        F: for<'a> Fn(&'a mut Transaction) -> BoxFuture<'a, Result<T, StashError>>,
     {
-        let mut tx = self.conn.transaction().await?;
-        let current: i64 = tx.query_one("SELECT epoch FROM fence", &[]).await?.get(0);
-        if current != self.epoch {
+        // TODO(mjibson): Using mz_ore::retry here would be nice, but I did not find a
+        // way to get the `f` lifetimes correct, so roll our own.
```

Comment:
Can you use `ore::Retry` and then call `into_retry_stream`? That should give you a stream that you can call `next` on, and then you don't have to muck about with the closure.